### EDA

In [1]:
import pandas as pd

In [2]:
# load dataframe
df=pd.read_csv('data.csv',dtype='str')

In [3]:
df

ID.format  \
0     The published format which was employed   
1                            SNAP 2015 SHIFT    
2                            SNAP 2015 SHIFT    
3                            SNAP 2015 SHIFT    
4                            SNAP 2015 SHIFT    
...                                       ...   
6808                         SNAP 2015 SHIFT    
6809                         SNAP 2015 SHIFT    
6810                         SNAP 2015 SHIFT    
6811                         SNAP 2015 SHIFT    
6812                         SNAP 2015 SHIFT    

                            ID.completed  \
0     Case completed in Snap Interviewer   
1                              completed   
2                              completed   
3                              completed   
4                              completed   
...                                  ...   
6808                           completed   
6809                           completed   
6810                           completed   
6811                           completed   
6812                           completed   

                                            ID.language  \
0     What language would you like to complete the s...   
1                                               Spanish   
2                                               English   
3                                               English   
4                                               English   
...                                                 ...   
6808                                            English   
6809                                            English   
6810                                            English   
6811                                            English   
6812                                            English   

                     ID.site            ID.date                ID.start  \
0     Questionnaire location  Date of interview  Time interview started   
1                        NaN         06/14/2019                16:46:41   
2                        NaN         06/14/2019                17:10:02   
3                        NaN         06/14/2019                17:15:21   
4                        NaN         06/14/2019                16:51:28   
...                      ...                ...                     ...   
6808                     NaN         06/14/2019                16:51:46   
6809                     NaN         06/14/2019                16:54:50   
6810                     NaN         06/14/2019                16:58:30   
6811                     NaN         06/14/2019                15:10:17   
6812                     NaN         06/14/2019                17:11:57   

                        ID.endDate                ID.end  \
0     Completion date of interview  Time interview ended   
1                       06/14/2019              17:36:11   
2                       06/14/2019              17:38:25   
3                       06/14/2019              17:39:13   
4                       06/14/2019              17:46:58   
...                            ...                   ...   
6808                    06/14/2019              17:07:47   
6809                    06/14/2019              17:08:17   
6810                    06/14/2019              17:19:23   
6811                    06/14/2019              17:21:02   
6812                    06/14/2019              17:27:22   

                    ID.time  \
0     Duration of interview   
1                      49.5   
2                     28.38   
3                     23.87   
4                      55.5   
...                     ...   
6808                  16.02   
6809                  13.45   
6810                  20.88   
6811                 130.75   
6812                  15.42   

                                                     Q1  ...  \
0     Which, if any, of the following degrees are yo...  ...   
1                       Doctorate degree (PhD/DPhil/MD)  ...   
2                     

The first row is the questions.

In [4]:
# check the size of the dataset
df.shape

(6813, 274)

The dataset has 6812 data points and 274 columns. There are a lot of columns so our main goal is to clean it up.

In [5]:
# check column names
list(df.columns)

['ID.format',
 'ID.completed',
 'ID.language',
 'ID.site',
 'ID.date',
 'ID.start',
 'ID.endDate',
 'ID.end',
 'ID.time',
 'Q1',
 'Q1.a',
 'Q2',
 'Q3',
 'Q3.a',
 'Q4',
 'Q5',
 'Q6',
 'Q6.a',
 'Q7',
 'Q7.a',
 'Q8',
 'Q8.a',
 'Q9',
 'Q9.a',
 'Q10',
 'Q10.a',
 'Q11',
 'Q11.a',
 'Q12:1',
 'Q12:2',
 'Q12:3',
 'Q12:4',
 'Q12:5',
 'Q12:6',
 'Q12:7',
 'Q12:8',
 'Q12:9',
 'Q12:10',
 'Q12:11',
 'Q12.a',
 'Q13',
 'Q14',
 'Q14.a',
 'Q15.a',
 'Q15.b',
 'Q15.c',
 'Q15.d',
 'Q15.e',
 'Q15.f',
 'Q15.g',
 'Q15.h',
 'Q15.i',
 'Q15.j',
 'Q15.k',
 'Q15.l',
 'Q15.m',
 'Q15.n',
 'Q16',
 'Q17',
 'Q17.a',
 'Q18.a',
 'Q19.a',
 'Q20',
 'Q21.a',
 'Q21.b',
 'Q21.c',
 'Q21.d',
 'Q21.e',
 'Q21.f',
 'Q21.g',
 'Q21.h',
 'Q21.i',
 'Q22.a',
 'Q22.b',
 'Q22.c',
 'Q22.d',
 'Q22.e',
 'Q22.f',
 'Q22.g',
 'Q22.h',
 'Q22.i',
 'Q23',
 'Q24',
 'Q25',
 'Q25.a',
 'Q26',
 'Q27.a',
 'Q27.b',
 'Q27.c',
 'Q27.d',
 'Q27.e',
 'Q27.f',
 'Q28',
 'Q29',
 'Q29.a',
 'Q30.a',
 'Q30.b',
 'Q30.c',
 'Q30.d',
 'Q30.e',
 'Q30.f',
 'Q31',
 'Q32:1

Questions start from the 10th column.

In [6]:
# print out a list of the questions
question_list=list(df.columns)[9:]
for q in question_list:
    print('********')
    print(q)
    print(df[q][0])

********
Q1
Which, if any, of the following degrees are you currently studying for?
********
Q1.a
If other, please specify
********
Q2
Hidden
********
Q3
Which was the most important reason you decided to enrol in a PhD programme?
********
Q3.a
If other, please specify
********
Q4
Are you studying in the country you grew up in?   
********
Q5
Where do you currently live?
********
Q6
Which region in Asia?
********
Q6.a
If other, please specify
********
Q7
Which country in Australasia?
********
Q7.a
If other, please specify
********
Q8
Which country in Africa?
********
Q8.a
If other, please specify
********
Q9
Which country in Europe?
********
Q9.a
If other, please specify
********
Q10
Which country in North or Central America?
********
Q10.a
If other, please specify
********
Q11
Which country in South America?
********
Q11.a
If other, please specify
********
Q12:1
What prompted you to study outside your country of upbringing? 
********
Q12:2
What prompted you to study outside your count

#### Target

Our target column is Q28

In [7]:
# check values for target
df['Q28'].value_counts()

No                                                                          4083
Yes                                                                         2429
Prefer not to say                                                            292
Have you ever sought help for anxiety or depression caused by PhD study?       1
Name: Q28, dtype: int64

In [8]:
df['Q28'].value_counts(normalize=True)

No                                                                          0.600000
Yes                                                                         0.356943
Prefer not to say                                                           0.042910
Have you ever sought help for anxiety or depression caused by PhD study?    0.000147
Name: Q28, dtype: float64

There are 4083 Nos and 2429 Yeses. There's a little bit of class imbalance, but not too bad.

In [9]:
# check null values
df['Q28'].isna().sum()

8

There are not many "prefer not to say"s or null values, so we can safely drop them.

In [10]:
df=df.dropna(subset=['Q28'])
df

ID.format  \
0     The published format which was employed   
1                            SNAP 2015 SHIFT    
2                            SNAP 2015 SHIFT    
3                            SNAP 2015 SHIFT    
4                            SNAP 2015 SHIFT    
...                                       ...   
6808                         SNAP 2015 SHIFT    
6809                         SNAP 2015 SHIFT    
6810                         SNAP 2015 SHIFT    
6811                         SNAP 2015 SHIFT    
6812                         SNAP 2015 SHIFT    

                            ID.completed  \
0     Case completed in Snap Interviewer   
1                              completed   
2                              completed   
3                              completed   
4                              completed   
...                                  ...   
6808                           completed   
6809                           completed   
6810                           completed   
6811                           completed   
6812                           completed   

                                            ID.language  \
0     What language would you like to complete the s...   
1                                               Spanish   
2                                               English   
3                                               English   
4                                               English   
...                                                 ...   
6808                                            English   
6809                                            English   
6810                                            English   
6811                                            English   
6812                                            English   

                     ID.site            ID.date                ID.start  \
0     Questionnaire location  Date of interview  Time interview started   
1                        NaN         06/14/2019                16:46:41   
2                        NaN         06/14/2019                17:10:02   
3                        NaN         06/14/2019                17:15:21   
4                        NaN         06/14/2019                16:51:28   
...                      ...                ...                     ...   
6808                     NaN         06/14/2019                16:51:46   
6809                     NaN         06/14/2019                16:54:50   
6810                     NaN         06/14/2019                16:58:30   
6811                     NaN         06/14/2019                15:10:17   
6812                     NaN         06/14/2019                17:11:57   

                        ID.endDate                ID.end  \
0     Completion date of interview  Time interview ended   
1                       06/14/2019              17:36:11   
2                       06/14/2019              17:38:25   
3                       06/14/2019              17:39:13   
4                       06/14/2019              17:46:58   
...                            ...                   ...   
6808                    06/14/2019              17:07:47   
6809                    06/14/2019              17:08:17   
6810                    06/14/2019              17:19:23   
6811                    06/14/2019              17:21:02   
6812                    06/14/2019              17:27:22   

                    ID.time  \
0     Duration of interview   
1                      49.5   
2                     28.38   
3                     23.87   
4                      55.5   
...                     ...   
6808                  16.02   
6809                  13.45   
6810                  20.88   
6811                 130.75   
6812                  15.42   

                                                     Q1  ...  \
0     Which, if any, of the following degrees are yo...  ...   
1                       Doctorate degree (PhD/DPhil/MD)  ...   
2                     

In [11]:
# check if those rows are sucessfully dropped
df['Q28'].value_counts()

No                                                                          4083
Yes                                                                         2429
Prefer not to say                                                            292
Have you ever sought help for anxiety or depression caused by PhD study?       1
Name: Q28, dtype: int64

In [12]:
# drop "Prefer not to say"s
df=df[df['Q28']!='Prefer not to say']

In [13]:
# check if those rows are sucessfully dropped 
df['Q28'].value_counts()

No                                                                          4083
Yes                                                                         2429
Have you ever sought help for anxiety or depression caused by PhD study?       1
Name: Q28, dtype: int64

#### Features

There are four different types of independent variables: binary ones which requires one hot encoding, multiclass ones that are not on a scale which require frequency encoding, multiclass ones that are on a scale which require further processing/value mapping, and numeric ones.

In [14]:
# make a list for each category of columns
bi_cols=[]
freq_cols=[]
map_cols=[]
num_cols=[]

The first 9 columns seem like statistics of the interview itself, not anything related to the person who got surveyed, expect 'ID.language', which indicates which language the surveyee completed the survey.

In [15]:
list(df.columns)

['ID.format',
 'ID.completed',
 'ID.language',
 'ID.site',
 'ID.date',
 'ID.start',
 'ID.endDate',
 'ID.end',
 'ID.time',
 'Q1',
 'Q1.a',
 'Q2',
 'Q3',
 'Q3.a',
 'Q4',
 'Q5',
 'Q6',
 'Q6.a',
 'Q7',
 'Q7.a',
 'Q8',
 'Q8.a',
 'Q9',
 'Q9.a',
 'Q10',
 'Q10.a',
 'Q11',
 'Q11.a',
 'Q12:1',
 'Q12:2',
 'Q12:3',
 'Q12:4',
 'Q12:5',
 'Q12:6',
 'Q12:7',
 'Q12:8',
 'Q12:9',
 'Q12:10',
 'Q12:11',
 'Q12.a',
 'Q13',
 'Q14',
 'Q14.a',
 'Q15.a',
 'Q15.b',
 'Q15.c',
 'Q15.d',
 'Q15.e',
 'Q15.f',
 'Q15.g',
 'Q15.h',
 'Q15.i',
 'Q15.j',
 'Q15.k',
 'Q15.l',
 'Q15.m',
 'Q15.n',
 'Q16',
 'Q17',
 'Q17.a',
 'Q18.a',
 'Q19.a',
 'Q20',
 'Q21.a',
 'Q21.b',
 'Q21.c',
 'Q21.d',
 'Q21.e',
 'Q21.f',
 'Q21.g',
 'Q21.h',
 'Q21.i',
 'Q22.a',
 'Q22.b',
 'Q22.c',
 'Q22.d',
 'Q22.e',
 'Q22.f',
 'Q22.g',
 'Q22.h',
 'Q22.i',
 'Q23',
 'Q24',
 'Q25',
 'Q25.a',
 'Q26',
 'Q27.a',
 'Q27.b',
 'Q27.c',
 'Q27.d',
 'Q27.e',
 'Q27.f',
 'Q28',
 'Q29',
 'Q29.a',
 'Q30.a',
 'Q30.b',
 'Q30.c',
 'Q30.d',
 'Q30.e',
 'Q30.f',
 'Q31',
 'Q32:1

In [16]:
df['ID.language'].value_counts()

English                                                    5196
Chinese (Simplified)                                        756
Spanish                                                     282
Portuguese                                                  157
French                                                      121
What language would you like to complete the survey in?       1
Name: ID.language, dtype: int64

In [17]:
# the rest of the columns
df=df.drop(columns=['ID.format',
 'ID.completed',
 'ID.site',
 'ID.date',
 'ID.start',
 'ID.endDate',
 'ID.end',
 'ID.time',])

In [18]:
df

ID.language  \
0     What language would you like to complete the s...   
1                                               Spanish   
2                                               English   
3                                               English   
5                                               English   
...                                                 ...   
6806                                             French   
6807                                            English   
6809                                            English   
6811                                            English   
6812                                            English   

                                                     Q1  \
0     Which, if any, of the following degrees are yo...   
1                       Doctorate degree (PhD/DPhil/MD)   
2                       Doctorate degree (PhD/DPhil/MD)   
3                       Doctorate degree (PhD/DPhil/MD)   
5                       Doctorate degree (PhD/DPhil/MD)   
...                                                 ...   
6806                    Doctorate degree (PhD/DPhil/MD)   
6807                    Doctorate degree (PhD/DPhil/MD)   
6809                    Doctorate degree (PhD/DPhil/MD)   
6811                    Doctorate degree (PhD/DPhil/MD)   
6812                    Doctorate degree (PhD/DPhil/MD)   

                          Q1.a      Q2  \
0     If other, please specify  Hidden   
1                          NaN     NaN   
2                          NaN     NaN   
3                          NaN     NaN   
5                          NaN     NaN   
...                        ...     ...   
6806                       NaN     NaN   
6807                       NaN     NaN   
6809                       NaN     NaN   
6811                       NaN     NaN   
6812                       NaN     NaN   

                                                     Q3  \
0     Which was the most important reason you decide...   
1              No job I want is available without a PhD   
2                   I want to pursue an academic career   
3             Personal interest in my subject of choice   
5              No job I want is available without a PhD   
...                                                 ...   
6806           No job I want is available without a PhD   
6807          I wanted to continue pursuing my research   
6809                I want to pursue an academic career   
6811          Personal interest in my subject of choice   
6812          Personal interest in my subject of choice   

                          Q3.a  \
0     If other, please specify   
1                          NaN   
2                          NaN   
3                          NaN   
5                          NaN   
...                        ...   
6806                       NaN   
6807                       NaN   
6809                       NaN   
6811                       NaN   
6812                       NaN   

                                                     Q4  \
0     Are you studying in the country you grew up in...   
1                                                   Yes   
2                                                    No   
3                                                   Yes   
5                                                   Yes   
...                                                 ...   
6806                                                Yes   
6807                                                 No   
6809                                                Yes   
6811                                                Yes   
6812                                                 No   

                                Q5                     Q6  \
0     Where do you currently live?  Which region in Asia?   
1         North or Central America                    NaN   
2         North or Central America                    NaN   
3         North or Central America                

In [19]:
def info(label):
    '''
    input string of column name
    output comprehensive analysis of the column
    '''
    print(f'column name: {label}')
    print('     ')
    print(f'Question: {df[label][0]}')
    print('     ')
    print(f'number of null values: {df[label].isna().sum()}')
    print('     ')
    print('Value Count:')
    
    print(df[label].value_counts())

In [20]:
# check Q1
info('Q1')

column name: Q1
     
Question: Which, if any, of the following degrees are you currently studying for?
     
number of null values: 0
     
Value Count:
Doctorate degree (PhD/DPhil/MD)                                            6294
Dual doctorate degree (MD-PhD, PhD-PhD or other combination)                218
Which, if any, of the following degrees are you currently studying for?       1
Name: Q1, dtype: int64


In [21]:
# there's Q1.a so we need to check that too to decide how to process this column
info('Q1.a')

column name: Q1.a
     
Question: If other, please specify
     
number of null values: 6512
     
Value Count:
If other, please specify    1
Name: Q1.a, dtype: int64


In [22]:
# Q1 and Q1.a tells us whether a student is doing dual degree or not
# drop Q1.a and map Q1 for 0 as not doing dual degree, 1 as doing dual degree
df=df.drop('Q1.a', axis=1)

In [23]:
dict_Q1={'Doctorate degree (PhD/DPhil/MD)': 0, 
         'Dual doctorate degree (MD-PhD, PhD-PhD or other combination)': 1, 
         'Which, if any, of the following degrees are you currently studying for?':'dual_degree'}

In [24]:
df['Q1']=df['Q1'].map(dict_Q1)

In [25]:
df['Q1'].value_counts()

0              6294
1               218
dual_degree       1
Name: Q1, dtype: int64

In [26]:
list(df.columns)

['ID.language',
 'Q1',
 'Q2',
 'Q3',
 'Q3.a',
 'Q4',
 'Q5',
 'Q6',
 'Q6.a',
 'Q7',
 'Q7.a',
 'Q8',
 'Q8.a',
 'Q9',
 'Q9.a',
 'Q10',
 'Q10.a',
 'Q11',
 'Q11.a',
 'Q12:1',
 'Q12:2',
 'Q12:3',
 'Q12:4',
 'Q12:5',
 'Q12:6',
 'Q12:7',
 'Q12:8',
 'Q12:9',
 'Q12:10',
 'Q12:11',
 'Q12.a',
 'Q13',
 'Q14',
 'Q14.a',
 'Q15.a',
 'Q15.b',
 'Q15.c',
 'Q15.d',
 'Q15.e',
 'Q15.f',
 'Q15.g',
 'Q15.h',
 'Q15.i',
 'Q15.j',
 'Q15.k',
 'Q15.l',
 'Q15.m',
 'Q15.n',
 'Q16',
 'Q17',
 'Q17.a',
 'Q18.a',
 'Q19.a',
 'Q20',
 'Q21.a',
 'Q21.b',
 'Q21.c',
 'Q21.d',
 'Q21.e',
 'Q21.f',
 'Q21.g',
 'Q21.h',
 'Q21.i',
 'Q22.a',
 'Q22.b',
 'Q22.c',
 'Q22.d',
 'Q22.e',
 'Q22.f',
 'Q22.g',
 'Q22.h',
 'Q22.i',
 'Q23',
 'Q24',
 'Q25',
 'Q25.a',
 'Q26',
 'Q27.a',
 'Q27.b',
 'Q27.c',
 'Q27.d',
 'Q27.e',
 'Q27.f',
 'Q28',
 'Q29',
 'Q29.a',
 'Q30.a',
 'Q30.b',
 'Q30.c',
 'Q30.d',
 'Q30.e',
 'Q30.f',
 'Q31',
 'Q32:1',
 'Q32:2',
 'Q32:3',
 'Q32:4',
 'Q32:5',
 'Q32:6',
 'Q32:7',
 'Q32.a',
 'Q33',
 'Q34',
 'Q35:1',
 'Q35:2',
 'Q35:

In [27]:
# check Q2
info('Q2')

column name: Q2
     
Question: Hidden
     
number of null values: 6512
     
Value Count:
Hidden    1
Name: Q2, dtype: int64


In [28]:
# Q2 is useless, drop it
df=df.drop('Q2', axis=1)

In [29]:
info('Q3')

column name: Q3
     
Question: Which was the most important reason you decided to enrol in a PhD programme?
     
number of null values: 0
     
Value Count:
I want to pursue an academic career                                             2505
Personal interest in my subject of choice                                       1396
I wanted to continue pursuing my research                                       1283
No job I want is available without a PhD                                         627
I want to pursue a non-academic career                                           422
Other, please specify                                                            155
I wanted to live in another country                                              124
Which was the most important reason you decided to enrol in a PhD programme?       1
Name: Q3, dtype: int64


In [30]:
# print out all the reasons
for reason in list(df[df['Q3.a'].notnull()]['Q3.a']):
    print(reason)
    

If other, please specify
Getting good job is difficult in India. In anticipation that one will get a good job... I joined MSc and later into PhD... But yet fruitful... Hence continue PhD
Bored at work
I wanted more job security than an RA level career would provide.
It was easier than getting a job in the United States where I moved with my spouse.
I watched an animate when I was a child, and I like to make what I saw in that animate come ture with my research, which requires hugh amount of knowledge and financial support.
Ability to do free research for another 3 year whilst having security in pay, work environment etc
I want to pursue my dream
wasn't decided yet on academic/non-academic career and wanted to keep my options open
Continue education
Love science and to learn new things
try to have different experience
Personal challenge, to become expert/especialized in a topic.
I wanted to continue pursuing my research and live in another country
in order to be employed n teach at Univ

Q3 requires frequency encoding.

In [32]:
info('Q13')

column name: Q13
     
Question: Do you have a job alongside your studies?
     
number of null values: 0
     
Value Count:
No                                           5249
Yes                                          1263
Do you have a job alongside your studies?       1
Name: Q13, dtype: int64
